File: fragment/nodes/sprite.py
from .drawable import Drawable
from pygame_render import Texture


class Sprite(Drawable):
    """Node that loads an image source to render on screen."""
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.sprite_image = None

    @property
    def image_source(self) -> str:
        return self.properties['image_source']

    @image_source.setter
    def image_source(self, source: str) -> None:
        self.properties['image_source'] = source
        self.sprite_image = self.window_manager.renderer.load_texture(self.properties['image_source'])

    def render(self) -> Texture:
        return self.sprite_image

File: fragment/nodes/sprite.js
import * as PIXI from 'pixi.js';

import { Drawable } from '/fragment/nodes/drawable.js';

export class Sprite extends Drawable {
    constructor(...args) {
        super(...args);
        this.spriteTexture = null;
    }

    get imageSource() {
        return this._properties['imageSource'];
    }

    set imageSource(source) {
        this._properties['imageSource'] = source;
        this.loadTexture(this._properties['imageSource']);
    }

    destroySelf() {
        if (this.spriteTexture) {
            this.spriteTexture.destroy();
            this.spriteTexture = null;
        }
        super.destroySelf();
    }

    /**
     * Load a texture from a source path asynchronously.
     * @param {string} source - The path to the image file.
     */
    async loadTexture(source) {
        try {
            if (this.spriteTexture) {
                this.spriteTexture.destroy();
                this.spriteTexture = null;
            }
            this.spriteTexture = await PIXI.Assets.load(source);
            if (!this.pixiSprite.destroyed) {
                this.pixiSprite.texture = this.spriteTexture;
            }
            else {
                this.spriteTexture.destroy();
            }
        } catch (error) {
            throw new Error(`Failed to load texture: ${source} ` + error);
        }
    }
}
